### PR TITLE
Fix communication object

### LINKED ISSUE #6: emc-system destructor does not work
To replicate: create an executable with an emc::io and return 0. 
You will see the following error message:
```
terminate called after throwing an instance of 'boost::wrapexcept<boost::lock_error>'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Fixes #6

### CODE REVIEW

**Comment by MatthijsBurgh** [comment]
Please perform an integrated test. I just tested this with the small example from https://github.com/tue-robotics/emc_system/issues/6#issuecomment-1261861952

**Comment by KdVos** [comment]
Tested your changes using the full example we provide to the students. Somehow 
the changes subtly breaks the full example. Instead of turning on the first object that is encountered, the robot stops moving and remain stationary. Will investigate what is happening.

File: src/communication.cpp [hunk 1]
```diff
@@ -19,34 +21,30 @@ Communication::Communication(std::string robot_name)
...
-    nh_bumper_f.setCallbackQueue(&bumper_f_cb_queue_);
-    sub_bumper_f_ = nh_bumper_f.subscribe<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, &Communication::bumperfCallback, this);
+    ros::SubscribeOptions bumper_f_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, boost::bind(&Communication::bumperfCallback, this, _1), ros::VoidPtr(), &odom_cb_queue_);
+    sub_bumper_f_ = nh.subscribe(bumper_f_sub_options);
 
-    ros::NodeHandle nh_bumper_b;
-    nh_bumper_b.setCallbackQueue(&bumper_b_cb_queue_);
-    sub_bumper_b_ = nh_bumper_b.subscribe<std_msgs::Bool>("/" + robot_name + "/base_b_bumper_sensor", 1, &Communication::bumperbCallback, this);
+    ros::SubscribeOptions bumper_b_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_b_bumper_sensor", 1, boost::bind(&Communication::bumperbCallback, this, _1), ros::VoidPtr(), &odom_cb_queue_);
+    sub_bumper_b_ = nh.subscribe(bumper_f_sub_options);
```
Comment (KdVos) [inline src/communication.cpp:40]:
```suggestion
    sub_bumper_b_ = nh.subscribe(bumper_b_sub_options);
```
;)

File: src/communication.cpp [hunk 2]
```diff
@@ -19,34 +21,30 @@ Communication::Communication(std::string robot_name)
...
-    ros::NodeHandle nh_bumper_f;
-    nh_bumper_f.setCallbackQueue(&bumper_f_cb_queue_);
-    sub_bumper_f_ = nh_bumper_f.subscribe<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, &Communication::bumperfCallback, this);
+    ros::SubscribeOptions bumper_f_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, boost::bind(&Communication::bumperfCallback, this, _1), ros::VoidPtr(), &odom_cb_queue_);
+    sub_bumper_f_ = nh.subscribe(bumper_f_sub_options);
 
-    ros::NodeHandle nh_bumper_b;
-    nh_bumper_b.setCallbackQueue(&bumper_b_cb_queue_);
-    sub_bumper_b_ = nh_bumper_b.subscribe<std_msgs::Bool>("/" + robot_name + "/base_b_bumper_sensor", 1, &Communication::bumperbCallback, this);
+    ros::SubscribeOptions bumper_b_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_b_bumper_sensor", 1, boost::bind(&Communication::bumperbCallback, this, _1), ros::VoidPtr(), &odom_cb_queue_);
```
Comment (KdVos) [inline src/communication.cpp:39]:
```suggestion
    ros::SubscribeOptions bumper_b_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_b_bumper_sensor", 1, boost::bind(&Communication::bumperbCallback, this, _1), ros::VoidPtr(), &bumper_b_cb_queue_);
```

File: src/communication.cpp [hunk 3]
```diff
@@ -19,34 +21,30 @@ Communication::Communication(std::string robot_name)
...
-    sub_ce_ = nh_ce.subscribe<emc_system::controlEffort>("/" + robot_name + "/controlEffort", 1, &Communication::controlEffortCallback, this);
-*/
+    ros::SubscribeOptions ce_sub_options = ros::SubscribeOptions::create<emc_system::controlEffort>("/" + robot_name + "/controlEffort", 1, boost::bind(&Communication::controlEffortCallback, this, _1), ros::VoidPtr(), &ce_cb_queue_);
+    sub_ce_ = nh.subscribe(ce_sub_options);
+    */
 
-    ros::NodeHandle nh_bumper_f;
-    nh_bumper_f.setCallbackQueue(&bumper_f_cb_queue_);
-    sub_bumper_f_ = nh_bumper_f.subscribe<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, &Communication::bumperfCallback, this);
+    ros::SubscribeOptions bumper_f_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, boost::bind(&Communication::bumperfCallback, this, _1), ros::VoidPtr(), &odom_cb_queue_);
```
Comment (KdVos) [inline src/communication.cpp:36]:
```suggestion
    ros::SubscribeOptions bumper_f_sub_options = ros::SubscribeOptions::create<std_msgs::Bool>("/" + robot_name + "/base_f_bumper_sensor", 1, boost::bind(&Communication::bumperfCallback, this, _1), ros::VoidPtr(), &bumper_f_cb_queue_);
```

**Review by KdVos — CHANGES_REQUESTED** [review]

**Comment by KdVos** [comment]
Apart from these minor oversights it seems to work as expected. Thanks!

File: src/communication.cpp (hunk 1 above)
Comment (MatthijsBurgh) [reply src/communication.cpp:40]: This is why you should review code.

**Review by KdVos — APPROVED** [review]

**Pull request merged**